src/Strategy/ManifestStrategy.php

```diff
@@ -0,0 +1,178 @@
+    {
+        $this->localVersion = $localVersion;
+        $this->manifestUrl = $manifestUrl;
+        $this->allowMajor = $allowMajor;
+        $this->allowUnstable = $allowUnstable;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCurrentLocalVersion(Updater $updater)
+    {
+        return $this->localVersion;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function download(Updater $updater)
+    {
+        $version = $this->getCurrentRemoteVersion($updater);
+        if ($version === false) {
+            throw new \RuntimeException('No remote versions found');
+        }
+        $versionInfo = $this->getAvailableVersions();
+        if (!isset($versionInfo[$version])) {
+            throw new \RuntimeException(sprintf('Failed to find manifest item for version %s', $version));
+        }
+
+        $fileContents = file_get_contents($versionInfo[$version]['url']);
+        if ($fileContents === false) {
+            throw new HttpRequestException(sprintf('Failed to download file from URL: %s', $versionInfo[$version]['url']));
+        }
+
+        $tmpFilename = $updater->getTempPharFile();
+        if (file_put_contents($tmpFilename, $fileContents) === false) {
+            throw new \RuntimeException(sprintf('Failed to write file: %s', $tmpFilename));
+        }
+
+        $tmpSha = sha1_file($tmpFilename);
+        if ($tmpSha !== $versionInfo[$version]['sha1']) {
+            unlink($tmpFilename);
+            throw new \RuntimeException(
+                sprintf(
+                    'SHA-1 verification failed: expected %s, actual %s',
+                    $versionInfo[$version]['sha1'],
+                    $tmpSha
+                )
+            );
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCurrentRemoteVersion(Updater $updater)
+    {
+        $versions = array_keys($this->getAvailableVersions());
+        if (!$this->allowMajor) {
+            $versions = $this->filterByLocalMajorVersion($versions);
+        }
+
+        $versionParser = new VersionParser($versions);
+
+        $mostRecent = $versionParser->getMostRecentStable();
+
+        // Look for unstable updates if explicitly allowed, or if the local
+        // version is already unstable and there is no new stable version.
+        if ($this->allowUnstable || ($versionParser->isUnstable($this->localVersion) && version_compare($mostRecent, $this->localVersion, '<'))) {
+            $mostRecent = $versionParser->getMostRecentAll();
+        }
+
+        return version_compare($mostRecent, $this->localVersion, '>') ? $mostRecent : false;
+    }
+
+    /**
+     * Get available versions to update to.
+     *
+     * @return array
+     *   An array keyed by the version name, whose elements are arrays
+     *   containing version information ('name', 'sha1', and 'url').
+     */
+    private function getAvailableVersions()
+    {
+        if (!isset($this->availableVersions)) {
```

Comment:
when possible, I would recommend an early return to avoid too much indentation :)

```php
if (isset($this->availableVersions)) {
    return $this->availableVersions;
}

// Keep going
```

Applies to other places as well